tests/unit/s2n_self_talk_alpn_test.c

```diff
@@ -13,413 +13,130 @@
-    EXPECT_EQUAL(waitpid(-1, &status, 0), pid);
-    EXPECT_EQUAL(status, 0);
-
-    /* Test a lower preferred matching ALPN request */
-    /* Create a pipe */
-    EXPECT_SUCCESS(pipe(server_to_client));
-    EXPECT_SUCCESS(pipe(client_to_server));
-
-    /* Create a child process */
-    pid = fork();
-    if (pid == 0) {
-        /* This is the child process, close the read end of the pipe */
-        EXPECT_SUCCESS(close(client_to_server[0]));
-        EXPECT_SUCCESS(close(server_to_client[1]));
-
-        /* Client only advertises our second choice, so we should negotiate it */
-        mock_client(client_to_server[1], server_to_client[0], &protocols[1], 1, protocols[1]);
-    }
-
-    /* This is the parent */
-    EXPECT_SUCCESS(close(client_to_server[1]));
-    EXPECT_SUCCESS(close(server_to_client[0]));
-
-    EXPECT_NOT_NULL(conn = s2n_connection_new(S2N_SERVER));
-    conn->server_protocol_version = S2N_TLS12;
-    conn->client_protocol_version = S2N_TLS12;
-    conn->actual_protocol_version = S2N_TLS12;
-    EXPECT_SUCCESS(s2n_connection_set_config(conn, config));
-
-    /* Set up the connection to read from the fd */
-    EXPECT_SUCCESS(s2n_connection_set_read_fd(conn, client_to_server[0]));
-    EXPECT_SUCCESS(s2n_connection_set_write_fd(conn, server_to_client[1]));
-
-    /* Negotiate the handshake. */
-
-    EXPECT_SUCCESS(s2n_negotiate(conn, &blocked));
-
-    for (int i = 1; i < 0xffff; i += 100) {
-        char *ptr = buffer;
-        int size = i;
-
-        do {
-            int bytes_read = 0;
-            EXPECT_SUCCESS(bytes_read = s2n_recv(conn, ptr, size, &blocked));
-
-            size -= bytes_read;
-            ptr += bytes_read;
-        } while (size);
-
-        for (int j = 0; j < i; j++) {
-            EXPECT_EQUAL(buffer[j], 33);
+    DEFER_CLEANUP(struct s2n_config *server_config = s2n_config_new(), s2n_config_ptr_free);
+    EXPECT_NOT_NULL(server_config);
+    EXPECT_SUCCESS(s2n_config_set_protocol_preferences(server_config, protocols,
+            s2n_array_len(protocols)));
+    EXPECT_SUCCESS(s2n_config_set_cipher_preferences(server_config, "default"));
+
+    DEFER_CLEANUP(struct s2n_cert_chain_and_key *chain_and_key = NULL,
+            s2n_cert_chain_and_key_ptr_free);
+    EXPECT_SUCCESS(s2n_test_cert_chain_and_key_new(&chain_and_key, S2N_DEFAULT_TEST_CERT_CHAIN,
+            S2N_DEFAULT_TEST_PRIVATE_KEY));
+    EXPECT_SUCCESS(s2n_config_add_cert_chain_and_key_to_store(server_config, chain_and_key));
+
+    struct {
+        int client_protocol_count;
+        const char *const *client_protocols;
+        const char *expected_protocol;
+    } test_cases[] = {
+        /* Test no client ALPN request */
+        { .client_protocol_count = 0, .client_protocols = NULL, .expected_protocol = NULL },
+
+        /* Test a matching ALPN request */
+        { .client_protocol_count = s2n_array_len(protocols),
+                .client_protocols = protocols,
+                .expected_protocol = protocols[0] },
```

Comment:
These might look a little less weird like this. Not sure if clang-format is forcing you to do this, though.
```suggestion
        { 
            .client_protocol_count = s2n_array_len(protocols),
            .client_protocols = protocols,
            .expected_protocol = protocols[0]
        },
```